js/private/js_binary.bzl

```diff
@@ -479,44 +478,31 @@ def _create_launcher(ctx, log_prefix_rule_set, log_prefix_rule, fixed_args = [],
     bash_launcher, toolchain_files = _bash_launcher(ctx, entry_point_path, log_prefix_rule_set, log_prefix_rule, fixed_args, fixed_env, is_windows, use_legacy_node_patches)
     launcher = create_windows_native_launcher_script(ctx, bash_launcher) if is_windows else bash_launcher
 
-    files = [bash_launcher] + toolchain_files
-    if ctx.attr.copy_data_to_bin:
-        if entry_point not in ctx.files.no_copy_to_bin:
-            files.append(copy_file_to_bin_action(ctx, entry_point))
-        else:
-            files.append(entry_point)
-        for d in ctx.files.data:
-            if d not in ctx.files.no_copy_to_bin:
-                files.append(copy_file_to_bin_action(ctx, d))
-            else:
-                files.append(d)
-    else:
-        files.append(entry_point)
-        files.extend(ctx.files.data)
+    launcher_files = [bash_launcher] + toolchain_files
     if use_legacy_node_patches:
-        files.extend(ctx.files._node_patches_legacy_files + [ctx.file._node_patches_legacy])
+        launcher_files.extend(ctx.files._node_patches_legacy_files + [ctx.file._node_patches_legacy])
     else:
-        files.extend(ctx.files._node_patches_files + [ctx.file._node_patches])
-    files.extend(ctx.toolchains["@rules_nodejs//nodejs:toolchain_type"].nodeinfo.tool_files)
+        launcher_files.extend(ctx.files._node_patches_files + [ctx.file._node_patches])
+    launcher_files.extend(ctx.toolchains["@rules_nodejs//nodejs:toolchain_type"].nodeinfo.tool_files)
     if ctx.attr.include_npm:
-        files.extend(ctx.toolchains["@rules_nodejs//nodejs:toolchain_type"].nodeinfo.npm_files)
-
-    runfiles = ctx.runfiles(
-        files = files,
-        transitive_files = gather_files_from_js_providers(
-            targets = ctx.attr.data,
-            include_transitive_sources = ctx.attr.include_transitive_sources,
-            include_declarations = ctx.attr.include_declarations,
-            include_npm_linked_packages = ctx.attr.include_npm_linked_packages,
-        ),
-    ).merge_all([
-        target[DefaultInfo].default_runfiles
-        for target in ctx.attr.data
-    ])
+        launcher_files.extend(ctx.toolchains["@rules_nodejs//nodejs:toolchain_type"].nodeinfo.npm_files)
+
+    runfiles = gather_runfiles(
+        ctx = ctx,
+        sources = [],
+        data = ctx.attr.data,
+        files = [entry_point] + ctx.files.data,
+        deps = [],
+        copy_files_to_bin = ctx.attr.copy_data_to_bin,
+        no_copy_to_bin = ctx.files.no_copy_to_bin,
+        include_transitive_sources = ctx.attr.include_transitive_sources,
+        include_declarations = ctx.attr.include_declarations,
+        include_npm_linked_packages = ctx.attr.include_npm_linked_packages,
+    )
```

Comment:
nit: you could do
```
runfiles = gather_runfiles(...).merge(ctx.runfiles(files = launcher_files))
```
instead so that
```
runfiles = runfiles,
```
can stay clean below